src/ripple/overlay/MessageScheduler.h

```diff
@@ -0,0 +1,496 @@
+#ifndef RIPPLE_OVERLAY_MESSAGESCHEDULER_H_INCLUDED
+#define RIPPLE_OVERLAY_MESSAGESCHEDULER_H_INCLUDED
+
+#include <ripple/basics/chrono.h>
+#include <ripple/basics/random.h>
+#include <ripple/beast/utility/Journal.h>
+#include <ripple/overlay/Peer.h>
+#include <ripple/overlay/impl/ProtocolMessage.h>
+#include <ripple.pb.h>
+
+#include <boost/asio/basic_waitable_timer.hpp>
+#include <boost/asio/io_service.hpp>
+#include <google/protobuf/message.h>
+
+#include <atomic>
+#include <cassert>
+#include <chrono>
+#include <cstdint>
+#include <functional>
+#include <iterator>
+#include <memory>
+#include <mutex>
+#include <ostream>
+#include <random>
+
+namespace ripple {
+
+/**
+ * We must hold idle peers by `std::weak_ptr` so that they can destruct upon
+ * disconnect.
+ * When we remove channels in `remove`, we need to find them by peer ID,
+ * because `std::weak_ptr` is not equality-comparable with anything,
+ * but we do not want to lock the `std::weak_ptr` to get that ID,
+ * so we copy it.
+ */
+struct Channel
+{
+    Peer::id_t peerId;
+    std::weak_ptr<Peer> peer;
+};
+
+inline Channel
+toChannel(std::shared_ptr<Peer> peer)
+{
+    return {peer->id(), peer};
+}
+
+class MessageScheduler
+{
+public:
+    enum class FailureCode {
+        // The request timed out.
+        TIMEOUT,
+        // The peer disconnected.
+        DISCONNECT,
+        // The message scheduler is shutting down.
+        SHUTDOWN,
+    };
+
+    using RequestId = uint32_t;
+
+    class Offer;
+
+    class Sender;
+
+    // A receiver has methods for success and failure callbacks.
+    // We package them together in the same object because they may share data,
+    // and we want to make it easy to control the lifetime of that data.
+    class Receiver;
+
+    // The `request_cookie` field on a `TMLedgerData` message represents
+    // a peer ID, which is a low number starting at 1 and believed to never
+    // exceed 300. By using a non-overlapping set of numbers for our request
+    // IDs, `PeerImp` can dispatch incoming responses `TMLedgerData` messages to
+    // the correct receiver. Similarly for the `seq` field on
+    // `TMGetObjectByHash` messages.
+    static constexpr RequestId MINIMUM_REQUEST_ID = 1 << 24;
+
+private:
+    using Clock = std::chrono::steady_clock;
+    using Timer = boost::asio::basic_waitable_timer<Clock>;
+    using Time = std::chrono::time_point<Clock>;
+
+    // An in-flight request.
+    struct Request
+    {
+        RequestId id;
+        Channel channel;
+        Receiver* receiver;
+        Timer timer;
+        Time sent;
+    };
+
+    boost::asio::io_service& io_service_;
+    beast::Journal journal_;
+
+    // Offers are negotiations between peers and senders.
+    // This mutex must be locked when handling either set.
+    std::mutex offersMutex_;
+    std::vector<Channel> channels_;
+    // TODO: Use a priority queue.
+    std::vector<Sender*> senders_;
```

Comment:
I don't love the lifetime management where the message scheduler is holds non-owning pointers. I'm used to asio, where objects are kept alive with `shared_ptr`s. I haven't looked at how the message scheduler is used from clients, but non-owning pointers gives me pause.